### PR TITLE
render fonts with pango, add global `font_family` config option

### PR DESCRIPTION
**Pull request by rtgiskard**
#### Describe your PR, what does it fix/add?

This resolves: https://github.com/hyprwm/Hyprland/issues/6073

#### Is there anything you want to mention? (unchecked code, possible bugs, found problems, breaking compatibility, etc.)

1. It might be better to allow dynamic apply `font_family` without restarting Hyprland
2. the `count_missing_glyphs()` may not behave as I think, basically works, but not for sure

#### Is it ready for merging, or does it need work?

Likely to be ready

### CODE REVIEW

File: src/debug/HyprDebugOverlay.cpp [hunk 1]
```diff
@@ -105,23 +102,49 @@ int CHyprMonitorDebugOverlay::draw(int offset) {
...
-    const float FPS      = 1.f / (avgFrametime / 1000.f); // frametimes are in ms
-    const float idealFPS = m_dLastFrametimes.size();
-
-    cairo_select_font_face(g_pDebugOverlay->m_pCairo, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
-
-    cairo_set_font_size(g_pDebugOverlay->m_pCairo, 10);
+    const float           FPS      = 1.f / (avgFrametime / 1000.f); // frametimes are in ms
+    const float           idealFPS = m_dLastFrametimes.size();
+
+    static auto           font_family = CConfigValue<std::string>("misc:font_family");
```
Comment (vaxerski) [inline src/debug/HyprDebugOverlay.cpp:108]: no snake

File: src/debug/HyprDebugOverlay.cpp [hunk 2]
```diff
@@ -130,57 +153,35 @@ int CHyprMonitorDebugOverlay::draw(int offset) {
...
-    cairo_text_extents(g_pDebugOverlay->m_pCairo, text.c_str(), &cairoExtents);
-    if (cairoExtents.width > maxX)
-        maxX = cairoExtents.width;
+    showText(text.c_str(), 10);
+
+    pango_font_description_free(pangoFD);
+    g_object_unref(layoutText);
 
-    yOffset += 11;
+    double pos_x = 0, pos_y = 0;
```
Comment (vaxerski) [inline src/debug/HyprDebugOverlay.cpp:176]: no snake

File: src/hyprerror/HyprError.cpp [hunk 3]
```diff
@@ -94,12 +95,19 @@ void CHyprError::createQueued() {
 
     // draw the text with a common font
     const CColor textColor = m_cQueued.r + m_cQueued.g + m_cQueued.b < 0.2f ? CColor(1.0, 1.0, 1.0, 1.0) : CColor(0, 0, 0, 1.0);
-
-    cairo_select_font_face(CAIRO, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
-    cairo_set_font_size(CAIRO, FONTSIZE);
     cairo_set_source_rgba(CAIRO, textColor.r, textColor.g, textColor.b, textColor.a);
 
-    float yoffset     = TOPBAR ? FONTSIZE : Y - PAD + FONTSIZE;
+    static auto           font_family = CConfigValue<std::string>("misc:font_family");
```
Comment (vaxerski) [inline src/hyprerror/HyprError.cpp:100]: no snake

File: src/debug/HyprNotificationOverlay.cpp [hunk 4]
```diff
@@ -81,25 +70,25 @@ CBox CHyprNotificationOverlay::drawNotifications(CMonitor* pMonitor) {
...
+    auto                  iconBackendID = get_iconBackendID(*fontFamily);
+
+    PangoLayout*          layoutIcon = pango_cairo_create_layout(m_pCairo);
+    PangoLayout*          layoutText = pango_cairo_create_layout(m_pCairo);
+    PangoFontDescription* pangoFD    = pango_font_description_new();
 
-    cairo_select_font_face(m_pCairo, "Sans", CAIRO_FONT_SLANT_NORMAL, CAIRO_FONT_WEIGHT_NORMAL);
+    pango_font_description_set_family_static(pangoFD, (*fontFamily).c_str());
+    pango_font_description_set_style(pangoFD, PANGO_STYLE_NORMAL);
+    pango_font_description_set_weight(pangoFD, PANGO_WEIGHT_NORMAL);
```
Comment (rtgiskard) [inline src/debug/HyprNotificationOverlay.cpp:86]: Here I'm trying to get the latest `font_family` settings and apply it to the new draw, however it does not seem to use the configured font, even with a restart of Hyprland. As As it's setup just like `CHyprError::createQueued()`, which works well for dynamic font update, but this does not work, I can't figure it out, would you like to have look at it? the single commit: 260fe70960aa27412d555715024e14f4cf7679e6

Comment (vaxerski) [reply src/debug/HyprNotificationOverlay.cpp:86]: I don't think I've ever used that func to set a font family, maybe check out how hyprlock renders text? it uses pango too.

File: src/debug/HyprNotificationOverlay.cpp [hunk 5]
```diff
@@ -81,25 +70,25 @@ CBox CHyprNotificationOverlay::drawNotifications(CMonitor* pMonitor) {
...
     float                 maxWidth = 0;
 
-    const auto            SCALE = pMonitor->scale;
-
+    const auto            SCALE   = pMonitor->scale;
     const auto            MONSIZE = pMonitor->vecTransformedSize;
 
-    cairo_text_extents_t  cairoExtents;
-    int                   iconW = 0, iconH = 0;
+    static auto           fontFamily    = CConfigValue<std::string>("misc:font_family");
```
[the remark this answers is not in the document]
Comment (rtgiskard) [reply src/debug/HyprNotificationOverlay.cpp:76]: Well, this confirmed to be the issue related to `pango_font_description_set_family_static()`, which use a reference to the original font family string parsed from config, and I've assumed the parsed string would be there through the function call. However, this does not seem to be true,

**Comment by vaxerski** [comment]
tag me when it works for review

**Comment by rtgiskard** [comment]
Now it's all ready

**Comment by rtgiskard** [comment]
@vaxerski Ready for review

File: src/debug/HyprNotificationOverlay.cpp [hunk 6]
```diff
@@ -1,6 +1,20 @@
+#include <numeric>
+#include <pango/pangocairo.h>
 #include "HyprNotificationOverlay.hpp"
 #include "../Compositor.hpp"
-#include <pango/pangocairo.h>
+#include "../config/ConfigValue.hpp"
+#include "../helpers/FontUtils.hpp"
+
+eIconBackend get_iconBackendID(const std::string& fontFamily) {
```
Comment (vaxerski) [inline src/debug/HyprNotificationOverlay.cpp:8]: what the fuck is this naming convention lmao

iconBackendFromFont

File: src/helpers/FontUtils.cpp [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#include <pango/pangocairo.h>
+
+#include "FontUtils.hpp"
+
+int count_missing_glyphs(const std::string& text, const std::string& fontFamily) {
```
Comment (vaxerski) [inline src/helpers/FontUtils.cpp:5]: no snake

countMissingGlyphs

File: src/helpers/FontUtils.hpp [hunk 8]
```diff
@@ -0,0 +1,5 @@
+#pragma once
+
+#include <string>
+
+int count_missing_glyphs(const std::string& text, const std::string& fontFamily);
```
Comment (vaxerski) [inline src/helpers/FontUtils.hpp:5]: I'd avoid global funcs tbh. Why not `namespace FontUtils`?

**Review by vaxerski — COMMENTED** [review]
rest ok

File: src/debug/HyprNotificationOverlay.cpp (hunk 6 above)
Comment (rtgiskard) [reply src/debug/HyprNotificationOverlay.cpp:8]: 🤒 my fault

File: src/helpers/FontUtils.hpp (hunk 8 above)
Comment (rtgiskard) [reply src/helpers/FontUtils.hpp:5]: Let me say, should we just write an inline function, and reuse the cairo context, pango layout and font description in `drawNotifications()`, as it's called very frequent, the `iconBackendFromFont()` and `countMissingGlyphs()` are not likely to be reused anywhere else.

**Review by vaxerski — DISMISSED** [review]
lgtm

**Comment by rtgiskard** [comment]
wait, my bad ..

**Comment by vaxerski** [comment]
lol

**Comment by rtgiskard** [comment]
> lol

🫣 should be ready

**Review by vaxerski — DISMISSED** [review]
ok

**Comment by vaxerski** [comment]
needs a wiki mr

**Comment by rtgiskard** [comment]
> needs a wiki mr

Here you are: https://github.com/hyprwm/hyprland-wiki/pull/658

**Review by vaxerski — APPROVED** [review]
life is life

**Pull request merged**